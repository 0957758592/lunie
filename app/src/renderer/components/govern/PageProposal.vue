<template lang="pug">
tm-page(data-title='Proposal')
  div(slot="menu"): vm-tool-bar
    a(@click="commentOnProposal(proposal.proposal_id)"): i.material-icons comment
    a(@click="proposalIsSpam(proposal.proposal_id)"): i.material-icons error

  tm-part(v-if="proposal.type === 'Text' && proposal.proposal_status === 'VotingPeriod'")
    div(slot='title') Proposed by #[router-link(:to="{ name: 'delegate', params: { delegate: proposal.proposer }}") {{ proposal.proposer }}]
    text-block(:content="proposal.title")

tm-part(title='Time to vote: 13D 23H 27M' v-if="proposal.proposal_status === 'VotingPeriod'"): tm-form-struct(:submit='confirmVote')
    field-vote(@click.native="vote('yes')" dt='Yes' :dd='yesPct'
      color='hsl(120,50%,35%)' :active="votePick === 'yes'" :results="voteVisible")

    field-vote(@click.native="vote('no')" dt='No' :dd="noPct"
      color='hsl(0,50%,35%)' :active="votePick === 'no'" :results="voteVisible")

    field-vote(@click.native="vote('no_with_veto')" dt='No with veto' :dd="noWithVetoPct"
      color='hsl(330,50%,35%)' :active="votePick === 'no_with_veto'" :results="voteVisible")

    field-vote(@click.native="vote('abstain')" dt='Abstain' :dd="abstainPct"
      color='hsl(0,0%,35%)' :active="votePick === 'abstain'" :results="voteVisible")

    div(slot='footer')
      tm-btn(theme='cosmos' type='button' @click.native="toggleVoteVisible" value='Toggle Results')
      tm-btn(theme='cosmos' type='submit' value='Confirm Vote')
</template>

<script>
import { mapGetters } from "vuex"
import { TmBtn, TmFormStruct, TmPage, TmPart } from "@tendermint/ui"
import FieldVote from "common/TmFieldVote"
import TextBlock from "common/TextBlock"
import VmToolBar from "common/VmToolBar"
export default {
  name: `page-proposal`,
  components: {
    TmBtn,
    FieldVote,
    TmFormStruct,
    TmPage,
    TmPart,
    TextBlock,
    VmToolBar
  },
  computed: {
    ...mapGetters([`proposals`]),
    proposal() {
      if (this.proposals) {
        return this.proposals[this.$route.params.proposal]
      } else {
        return {}
      }
    },
    totalVotes() {
      return (
        this.proposal.tally_result.yes +
        this.proposal.tally_result.no +
        this.proposal.tally_result.no_with_veto +
        this.proposal.tally_result.abstain
      )
    },
    yesPct() {
      return this.proposal.tally_result.yes / this.totalVotes
    },
    noPct() {
      return this.proposal.tally_result.no / this.totalVotes
    },
    noWithVetoPct() {
      return this.proposal.tally_result.no_with_veto / this.totalVotes
    },
    abstainPct() {
      return this.proposal.tally_result.abstain / this.totalVotes
    }
  },
  data: () => ({
    votePick: `abstain`,
    voteVisible: false
  }),
  methods: {
    vote(choice) {
      this.votePick = choice
    },
    confirmVote() {
      this.$store.commit(`notify`, {
        title: `Voted '${this.votePick}'`,
        body: `Your vote has been submitted.`
      })
    },
    commentOnProposal(proposalId) {
      this.$store.commit(`notify`, {
        title: `View Discussion Thread`,
        body: `TODO: Discuss ${proposalId} proposal on the forum.`
      })
    },
    proposalIsSpam(proposalId) {
      this.$store.commit(`notify`, {
        title: `Mark Proposal As Spam`,
        body: `TODO: Mark ${proposalId} proposal as a spam attack.`
      })
    },
    toggleVoteVisible() {
      this.voteVisible = !this.voteVisible
    }
  }
}
</script>
