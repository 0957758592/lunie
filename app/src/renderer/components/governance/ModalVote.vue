<template lang="pug">
  .modal-vote#modal-vote(v-click-outside="close")
    .modal-vote-header
      img.icon(class='modal-vote-atom' src="~assets/images/cosmos-logo.png")
      span.tm-modal-title Cast Vote
      .tm-modal-icon.tm-modal-close#closeBtn(@click="close()")
        i.material-icons close

    tm-form-group.modal-vote-form-group.options
      tm-btn#vote-yes(
        @click.native="voteYes"
        color="secondary"
        value="Yes"
        size="md")

      tm-btn#vote-no(
        @click.native="voteNo"
        color="secondary"
        value="No"
        size="md")

      tm-btn#vote-veto(
        @click.native="voteVeto"
        color="secondary"
        value="No With Veto"
        size="md")

      tm-btn#vote-abstain(
        @click.native="voteAbstain"
        color="secondary"
        value="Abstain"
        size="md")

    tm-form-group.modal-vote-form-group(
      field-id='checkbox'
    )

      tm-field#checkbox(
        type="checkbox"
        v-model="approve"
        )

    tm-form-group.modal-vote-form-group
    .modal-vote-footer
      tm-btn(
        @click.native="onVote"
        :disabled="$v.option.$invalid"
        color="primary"
        value="Vote"
        size="lg")
</template>

<script>
import { mapGetters } from "vuex"
import ClickOutside from "vue-click-outside"
import { required, helpers } from "vuelidate/lib/validators"
import Modal from "common/TmModal"
import { TmBtn, TmField, TmFormGroup, TmFormMsg } from "@tendermint/ui"

const isValid = option =>
  option === `Yes` ||
  option === `No` ||
  option === `No With Veto` ||
  option === `Abstain`

const checked = value => !!value

export default {
  name: `modal-vote`,
  props: [`proposalId`],
  components: {
    Modal,
    TmBtn,
    TmField,
    TmFormGroup,
    TmFormMsg
  },
  data: () => ({
    option: ``,
    approve: false
  }),
  validations() {
    return {
      option: {
        required,
        isValid
      },
      approve: {
        required,
        isValid
      }
    }
  },
  methods: {
    close() {
      this.$emit(`update:showModalVote`, false)
    },
    voteYes() {
      this.option = `yes`
    },
    voteNo() {
      this.option = `no`
    },
    voteVeto() {
      this.option = `no_with_veto`
    },
    voteAbstain() {
      this.option = `abstain`
    },
    onVote() {
      if (this.approve) {
        this.$emit(`submitVote`, {
          proposal_id: this.proposalId,
          option: this.option
        })
        this.close()
      }
    }
  },
  directives: {
    ClickOutside
  }
}
</script>

<style lang="stylus">
@import '~variables'

.modal-vote
  background var(--app-nav)
  display flex
  flex-direction column
  height 50%
  justify-content space-between
  left 50%
  padding 2rem
  position fixed
  top 50%
  width 40%
  z-index z(modal)

  &-header
    align-items center
    display flex

  &-form-group
    display block
    padding 0

    .options

  &-footer
    display flex
    justify-content flex-end

    button
      margin-left 1rem
</style>
