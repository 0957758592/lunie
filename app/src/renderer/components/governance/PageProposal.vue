<template lang="pug">
tm-page(data-title='Proposal')
  template(slot="menu-body"): tm-balance
  div(slot="menu"): tm-tool-bar
    router-link(to="/governance" exact): i.material-icons arrow_back

  tm-data-error(v-if="!proposal")

  template(v-else)
    .validator-profile__header.validator-profile__section.proposal
      .column.validator-profile__header__info
        .row.validator-profile__header__name
          .top.column
            div.validator-profile__status-and-title
              span.validator-profile__status(v-bind:class="status.color" v-tooltip.top="status.message")
              .validator-profile__header__name__title {{ proposal.title }}
          .column.validator-profile__header__actions
            tm-btn(v-if="status.button === 'vote'" value="Vote" color="primary")
            tm-btn(v-if="status.button === 'deposit'" value="Deposit" color="primary")
            tm-btn(v-if="!status.button" disabled value="Deposit / Vote" color="primary")

        .row.description
          p This {{ proposalType }} proposal ({{ `#` + proposal.proposal_id }}) was submitted at block {{ submitBlock }} and voting started at {{ voteBlock }}.

        .row.validator-profile__header__data.votes
          dl.colored_dl
            dt Deposit
            dd {{ proposal.total_deposit[0].amount + ` ` + proposal.total_deposit[0].denom }}
          .validator-profile__header__data__break
          dl.colored_dl
            dt Yes
            dd {{ proposal.tally_result.yes }} / {{ yesPercentage }}
          dl.colored_dl
            dt No
            dd {{ proposal.tally_result.no }} / {{ noPercentage }}
          dl.colored_dl
            dt No with Veto
            dd {{ proposal.tally_result.no_with_veto }} / {{ noWithVetoPercentage }}
          dl.colored_dl
            dt Abstain
            dd {{ proposal.tally_result.abstain }} / {{ abstainPercentage }}

    .validator-profile__details.validator-profile__section
      .column
        .row
          text-block(:content="proposal.description")
</template>

<script>
import num from "scripts/num"
import { TmBtn, TmPage, TmToolBar } from "@tendermint/ui"
import TmBalance from "common/TmBalance"
import FieldVote from "common/TmFieldVote"
import TextBlock from "common/TextBlock"
export default {
  name: `page-proposal`,
  props: [`proposal`],
  components: {
    TmBalance,
    TmBtn,
    FieldVote,
    TmToolBar,
    TmPage,
    TextBlock
  },
  computed: {
    proposalType() {
      return this.proposal.proposal_type.toLowerCase()
    },
    submitBlock() {
      return `#` + num.prettyInt(this.proposal.submit_block)
    },
    voteBlock() {
      if (this.proposal.submit_block === this.proposal.voting_start_block) {
        return `the same block`
      } else {
        return `block #` + num.prettyInt(this.proposal.voting_start_block)
      }
    },
    status() {
      switch (this.proposal.proposal_status) {
        case `Passed`:
          return {
            button: null,
            message: `This proposal has passed`,
            color: `green`
          }
          break
        case `Rejected`:
          return {
            button: null,
            message: `This proposal has been rejected and voting is closed`,
            color: `red`
          }
          break
        case `Active`:
          return {
            button: `vote`,
            message: `Voting for this proposal is open`,
            color: `blue`
          }
          break
        case `Pending`:
          return {
            button: `deposit`,
            message: `Deposits are open for this proposal`,
            color: `yellow`
          }
          break
        default:
          return {
            button: null,
            message: `There was an error determining the status of this proposal`,
            color: `grey`
          }
      }
    },
    totalVotes() {
      return (
        this.proposal.tally_result.yes +
        this.proposal.tally_result.no +
        this.proposal.tally_result.no_with_veto +
        this.proposal.tally_result.abstain
      )
    },
    yesPercentage() {
      return num.percentInt(this.proposal.tally_result.yes / this.totalVotes)
    },
    noPercentage() {
      return num.percentInt(this.proposal.tally_result.no / this.totalVotes)
    },
    noWithVetoPercentage() {
      return num.percentInt(
        this.proposal.tally_result.no_with_veto / this.totalVotes
      )
    },
    abstainPercentage() {
      return num.percentInt(
        this.proposal.tally_result.abstain / this.totalVotes
      )
    }
  }
}
</script>
<style lang="stylus">
@require '~variables';
.proposal
  b
    color var(--bright)

  .validator-profile__status
    position relative
    left 0
    margin-right 0.5rem

  .description
    max-width 500px

  .votes
    padding-top 2rem

.proposal-id
  color var(--dim)
  font-size 14px
  margin 0
  font-weight 400
  padding-bottom 0.25rem

.text-block
  padding 0

.row b
  font-weight 500
</style>
