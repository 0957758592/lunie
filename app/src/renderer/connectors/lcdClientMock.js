'use strict'

let state = {
  keys: [{
    name: 'MOCK_ACCOUNT',
    password: '1234567890',
    address: 'DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B'
  }],
  accounts: {
    DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B: {
      coins: [{
        denom: 'mycoin',
        amount: 1000
      }]
    }
  },
  nonces: { DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B: 0 },
  txs: [{
    tx: {
      hash: 'x',
      inputs: [
        {
          coins: [{
            denom: 'jbcoins',
            amount: 1234
          }],
          sender: makeAddress()
        }
      ],
      outputs: [
        {
          coins: [{
            denom: 'jbcoins',
            amount: 1234
          }],
          receiver: 'DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B'
        }
      ]
    },
    height: 1
  }, {
    tx: {
      hash: 'y',
      inputs: [
        {
          coins: [{
            denom: 'fabocoins',
            amount: 1234
          }],
          sender: 'DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B'
        }
      ],
      outputs: [
        {
          coins: [{
            denom: 'fabocoins',
            amount: 1234
          }],
          receiver: makeAddress()
        }
      ]
    },
    height: 150
  }],
  stake: {
    DF096FDE8D380FA5B2AD20DB2962C82DDEA1ED9B: {
      '7A9D783CE542B23FA23DC7F101460879861205772606B4C3FAEAFBEDFB00E7BD': {
        PubKey: { data: '7A9D783CE542B23FA23DC7F101460879861205772606B4C3FAEAFBEDFB00E7BD' },
        Shares: 5
      }
    }
  },
  delegates: [
    {
      'address': '70705055A9FA5901735D0C3F0954501DDE667327',
      'pub_key': {
        'type': 'ed25519',
        'data': '88564A32500A120AA72CEFBCF5462E078E5DDB70B6431F59F778A8DC4DA719A4'
      },
      'voting_power': 14,
      'accum': 585,
      description: {
        description: 'Mr Mounty',
        moniker: 'mr_mounty',
        country: 'Canada'
      }
    },
    {
      'address': '760ACDE75EFC3DD0E4B2A6A3B96D91C05349EA31',
      'pub_key': {
        'type': 'ed25519',
        'data': '7A9D783CE542B23FA23DC7F101460879861205772606B4C3FAEAFBEDFB00E7BD'
      },
      'voting_power': 32,
      'accum': -1107,
      description: {
        description: 'Good Guy Greg',
        moniker: 'good_greg',
        country: 'USA'
      }
    },
    {
      'address': '77C26DF82654C5A5DDE5C6B7B27F3F06E9C223C0',
      'pub_key': {
        'type': 'ed25519',
        'data': '651E7B12B3C7234FB82B4417C59DCE30E4EA28F06AD0ACAEDFF05F013E463F10'
      },
      'voting_power': 19,
      'accum': 539,
      description: {
        description: 'Herr Schmidt',
        moniker: 'herr_schmidt',
        country: 'DE'
      }
    }
  ]
}

module.exports = {
  async lcdConnected () { return true },
  async sign ({ name, password, tx }) {
    return {
      'type': 'sigs/one',
      'data': {
        'tx': tx,
        'signature': {
          'Sig': {
            'type': 'ed25519',
            'data': '4631B03666EDF3D95E67A0700AD93741728E9466E51A8684F8637B935BB9A296ACE66256F2EDEC4F469FB9A940A4C1D30BB9170B5CA5292A6570A9F836DB7609'
          },
          'Pubkey': {
            'type': 'ed25519',
            'data': '81B11E717789600CC192B26F452A983DF13B985EE75ABD9DD9E68D7BA007A958'
          }
        }
      }
    }
  },
  async postTx (signedTx) {
    let outerTx = signedTx.data.tx.data.tx.data.tx
    let address = outerTx.data.signers[0].addr
    state.txs.push(outerTx.data.tx)
    state.nonces[address]++

    return {
      'check_tx': {
        'code': 0,
        'data': '',
        'log': '',
        'gas': '0',
        'fee': '0'
      },
      'deliver_tx': {
        'code': 0,
        'data': '',
        'log': '',
        'tags': []
      },
      'hash': '999ADECC2DE8C3AC2FD4F45E5E1081747BBE504A',
      'height': 0
    }
  },

  // keys
  async generateKey ({ name, password }) {
    let key = {
      name, password, address: makeAddress()
    }
    state.keys.push(key)
    return key
  },
  async listKeys () {
    return state.keys.map(k => ({
      name: k.name,
      address: k.address
    }))
  },
  async getKey (name) {
    return state.keys.find(k => k.name === name)
  },
  async updateKey (account, { name, password, new_passphrase }) { // eslint-disable-line camelcase
    let key = state.keys.find(k => k.name === name)
    if (key.password !== password) {
      throw new Error('Passwords do not match')
    }
    key.password = new_passphrase // eslint-disable-line camelcase
  },
  // axios handles DELETE requests different then other requests, we have to but the body in a config object with the prop data
  async deleteKey (account, { name, password }) {
    let key = state.keys.find(k => k.name === name)
    key
    if (key.password !== password) {
      throw new Error('Passwords do not match')
    }
    state.keys = state.keys.filter(k => k.name === name)
  },
  async recoverKey ({ name, password, seed_phrase }) { // eslint-disable-line camelcase
    let key = {
      name, password, address: makeAddress()
    }
    state.keys.push(key)
    return key
  },

  // coins
  async buildSend ({ sequence, from: { addr: fromAddr }, to: { addr: toAddr }, fees, amount }) {
    return {
      'type': 'sigs/one',
      'data': {
        'tx': {
          'type': 'chain/tx',
          'data': {
            'chain_id': 'mock-chain',
            'expires_at': 0,
            'tx': {
              'type': 'nonce',
              'data': {
                'sequence': sequence,
                'signers': [
                  {
                    'chain': '',
                    'app': 'sigs',
                    'addr': fromAddr
                  }
                ],
                'tx': {
                  'type': 'coin/send',
                  'data': {
                    'inputs': [
                      {
                        'address': {
                          'chain': '',
                          'app': 'sigs',
                          'addr': fromAddr
                        },
                        'coins': amount
                      }
                    ],
                    'outputs': [
                      {
                        'address': {
                          'chain': '',
                          'app': 'sigs',
                          'addr': toAddr
                        },
                        'coins': amount
                      }
                    ]
                  }
                }
              }
            }
          }
        },
        'signature': {
          'Sig': null,
          'Pubkey': null
        }
      }
    }
  },
  async queryAccount (address) {
    let account = state.accounts[address]
    if (account) {
      return { data: state.accounts[address] }
    }
    return { data: { coins: [] } }
  },
  async coinTxs (address) {
    return state.txs.filter(tx => {
      return tx.tx.inputs.find(input => input.sender === address) || tx.tx.outputs.find(output => output.receiver === address)
    })
  },

  // nonce
  async queryNonce (address) {
    return state.nonces[address]
  },

  // staking
  async candidate (pubkey) {
    return { data: state.delegates.find(delegate => delegate.pub_key.data === pubkey) }
  },
  async candidates () {
    return { data: state.delegates.map(({ pub_key }) => pub_key) } // eslint-disable-line camelcase
  },
  async buildDelegate ({ sequence, from, amount, pubkey }) {
    return {
      'type': 'sigs/one',
      'data': {
        'tx': {
          'type': 'chain/tx',
          'data': {
            'chain_id': 'mock-chain',
            'expires_at': 0,
            'tx': {
              'type': 'nonce',
              'data': {
                'sequence': sequence,
                'signers': [
                  {
                    'chain': '',
                    'app': 'sigs',
                    'addr': from.addr
                  }
                ],
                'tx': {
                  'type': 'stake/delegate',
                  'data': {
                    'pub_key': pubkey,
                    'amount': amount
                  }
                }
              }
            }
          }
        },
        'signature': {
          'Sig': null,
          'Pubkey': null
        }
      }
    }
  },
  async buildUnbond () {
    throw Error('Not implemented')
  },
  async bondingsByDelegator ([address, delegatePubkey]) {
    let stake = state.stake[address]
    if (stake && stake[delegatePubkey]) {
      return { data: stake[delegatePubkey] }
    }
    return {
      data: {
        PubKey: { data: delegatePubkey },
        Shares: 0
      }
    }
  }
}

function makeAddress () {
  var text = ''
  var possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

  for (var i = 0; i < 40; i++) { text += possible.charAt(Math.floor(Math.random() * possible.length)) }

  return text
}
