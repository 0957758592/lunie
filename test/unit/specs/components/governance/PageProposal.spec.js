"use strict"

import Vuelidate from "vuelidate"
import setup from "../../../helpers/vuex-setup"
import PageProposal from "renderer/components/governance/PageProposal"
import ModalVote from "renderer/components/governance/ModalVote"
import lcdClientMock from "renderer/connectors/lcdClientMock.js"
import { mount } from "@vue/test-utils"

let proposal = lcdClientMock.state.proposals[0]
let status = {
  button: null,
  message: `This proposal has passed`,
  color: `green`
}

describe(`PageProposal`, () => {
  describe(`with magic`, () => {
    let wrapper
    let { mount, localVue } = setup()
    localVue.use(Vuelidate)

    const $store = {
      commit: jest.fn(),
      dispatch: jest.fn()
    }

    beforeEach(() => {
      let instance = mount(PageProposal, {
        localVue,
        propsData: {
          proposal,
          status
        },
        $store
      })
      wrapper = instance.wrapper
      wrapper.update()
    })

    it(`has the expected html structure`, async () => {
      // after importing the @tendermint/ui components from modules
      // the perfect scroll plugin needs a $nextTick and a wrapper.update
      // to work properly in the tests (snapshots weren't matching)
      // this has occured across multiple tests
      await wrapper.vm.$nextTick()
      wrapper.update()
      expect(wrapper.vm.$el).toMatchSnapshot()
    })

    it(`should return the block number`, () => {
      expect(wrapper.vm.voteBlock).toBe(`block #135`)
    })

    it(`should return the end of the sentence`, () => {
      proposal.submit_block = `135`
      let { wrapper } = mount(PageProposal, {
        propsData: {
          proposal,
          status
        }
      })
      wrapper.update()
      expect(wrapper.vm.voteBlock).toBe(`the same block`)
    })

    describe(`Modal onVote`, () => {
      it(`enables voting if the proposal is Active`, () => {
        let status = { button: `vote` }
        wrapper.setProps({ status })

        let voteBtn = wrapper.find(`#vote-btn`)
        voteBtn.trigger(`click`)
        expect(wrapper.contains(ModalVote)).toEqual(true)
        expect(voteBtn.html()).not.toContain(`disabled="disabled"`)
      })

      it(`disables voting if the proposal is Pending deposits`, () => {
        let status = { button: `deposit` }
        wrapper.setProps({ status })
        expect(wrapper.find(`#vote-btn`).exists()).toEqual(false)
      })
    })
  })

  describe(`without magic`, () => {
    describe(`Vote`, () => {
      describe(`unit`, () => {
        it(`success`, async () => {
          const $store = {
            commit: jest.fn(),
            dispatch: jest.fn(),
            getters: {
              bondingDenom: `atom`,
              totalAtoms: 100,
              user: { atoms: 42 }
            }
          }

          const wrapper = mount(PageProposal, {
            mocks: { $store },
            propsData: {
              proposal: lcdClientMock.state.proposals[0],
              status: { message: `message` }
            }
          })

          await wrapper.vm.castVote({ option: `no_with_veto` })

          expect($store.dispatch.mock.calls).toEqual([
            [`submitVote`, { option: `no_with_veto`, proposalId: `1` }]
          ])

          expect($store.commit.mock.calls).toEqual([
            [
              `notify`,
              {
                body: `You have successfully voted no_with_veto on proposal #1`,
                title: `Successful vote!`
              }
            ]
          ])
        })

        it(`error`, async () => {
          const dispatch = jest.fn(() => {
            throw new Error(`one\ntwo\nthree\nfour\nfive\nsix\nseven`)
          })

          const $store = {
            commit: jest.fn(),
            dispatch,
            getters: {
              bondingDenom: `atom`,
              totalAtoms: 100,
              user: { atoms: 42 }
            }
          }

          const wrapper = mount(PageProposal, {
            mocks: { $store },
            propsData: {
              proposal: lcdClientMock.state.proposals[0],
              status: { message: `message` }
            }
          })

          await wrapper.vm.castVote({ option: `abstain` })

          expect($store.dispatch.mock.calls).toEqual([
            [`submitVote`, { option: `abstain`, proposalId: `1` }]
          ])

          expect($store.commit.mock.calls).toEqual([
            [
              `notifyError`,
              { body: `six`, title: `Error while voting on proposal #1` }
            ]
          ])
        })
      })
    })
  })
})
