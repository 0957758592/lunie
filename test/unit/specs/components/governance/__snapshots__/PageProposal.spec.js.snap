// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageProposal renders votes in HTML when voting is open 1`] = `
<tmpage-stub
  class="small"
  data-title="Proposal"
  hideheader="true"
  subtitle=""
  title=""
>
  <div
    class="page-profile__header page-profile__section proposal"
  >
    <div
      class="page-profile__header__info"
    >
      <span
        class="proposal-status pink"
      >
        Voting Period
      </span>
       
      <div
        class="page-profile__status-and-title"
      >
        <h2
          class="proposal-title"
        >
          VotingPeriod proposal
        </h2>
      </div>
    </div>
     
    <div
      class="button-container"
    >
      <tmbtn-stub
        color="primary"
        disabled="true"
        id="deposit-btn"
        value="Deposit"
      />
       
      <tmbtn-stub
        color="primary"
        id="vote-btn"
        value="Vote"
      />
    </div>
     
    <div
      class="row"
    >
      <div
        class="column"
      >
        <dl>
          <textblock-stub
            content="custom text proposal description"
          />
        </dl>
      </div>
    </div>
  </div>
   
  <div
    class="page-profile__section"
  >
    <div
      class="row row-condensed"
    >
      <!---->
    </div>
     
    <div
      class="row row-condensed"
    >
      <dl>
        <dt>
          Total Vote Count
        </dt>
         
        <dd>
          100,000,000% / 10,000
        </dd>
      </dl>
       
      <dl>
        <dt>
          Yes
        </dt>
         
        <dd>
          
            10% /
            1,000
          
        </dd>
      </dl>
       
      <dl>
        <dt>
          No
        </dt>
         
        <dd>
          
            20% /
            2,000
          
        </dd>
      </dl>
       
      <dl>
        <dt>
          No with Veto
        </dt>
         
        <dd>
          
            30% /
            3,000
          
        </dd>
      </dl>
       
      <dl>
        <dt>
          Abstain
        </dt>
         
        <dd>
          
            40% /
            4,000
          
        </dd>
      </dl>
    </div>
     
    <div
      class="row row-condensed"
    >
      <dl>
        <dt>
          Proposal ID
        </dt>
         
        <dd>
          2
        </dd>
      </dl>
       
      <dl>
        <dt>
          Submitted
        </dt>
         
        <dd>
          January 1st 1970, 00:00
        </dd>
      </dl>
       
      <dl>
        <dt>
          Voting Start Date
        </dt>
         
        <dd>
          January 3rd 1970, 00:00
        </dd>
      </dl>
       
      <dl>
        <dt>
          Voting End Date
        </dt>
         
        <dd>
          January 5th 1970, 00:00
        </dd>
      </dl>
    </div>
  </div>
   
  <modaldeposit-stub
    denom="STAKE"
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
   
  <modalvote-stub
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
</tmpage-stub>
`;

exports[`PageProposal should display proposal page if user has signed in 1`] = `
<tmpage-stub
  class="small"
  data-title="Proposal"
  hideheader="true"
  subtitle=""
  title=""
>
  <div
    class="page-profile__header page-profile__section proposal"
  >
    <div
      class="page-profile__header__info"
    >
      <span
        class="proposal-status pink"
      >
        Voting Period
      </span>
       
      <div
        class="page-profile__status-and-title"
      >
        <h2
          class="proposal-title"
        >
          VotingPeriod proposal
        </h2>
      </div>
    </div>
     
    <div
      class="button-container"
    >
      <tmbtn-stub
        color="primary"
        disabled="true"
        id="deposit-btn"
        value="Deposit"
      />
       
      <tmbtn-stub
        color="primary"
        id="vote-btn"
        value="Vote"
      />
    </div>
     
    <div
      class="row"
    >
      <div
        class="column"
      >
        <dl>
          <textblock-stub
            content="custom text proposal description"
          />
        </dl>
      </div>
    </div>
  </div>
   
  <div
    class="page-profile__section"
  >
    <div
      class="row row-condensed"
    >
      <!---->
    </div>
     
    <div
      class="row row-condensed"
    >
      <dl>
        <dt>
          Total Vote Count
        </dt>
         
        <dd>
          0% / 0
        </dd>
      </dl>
       
      <dl>
        <dt>
          Yes
        </dt>
         
        <dd>
          
            0% /
            0
          
        </dd>
      </dl>
       
      <dl>
        <dt>
          No
        </dt>
         
        <dd>
          
            0% /
            0
          
        </dd>
      </dl>
       
      <dl>
        <dt>
          No with Veto
        </dt>
         
        <dd>
          
            0% /
            0
          
        </dd>
      </dl>
       
      <dl>
        <dt>
          Abstain
        </dt>
         
        <dd>
          
            0% /
            0
          
        </dd>
      </dl>
    </div>
     
    <div
      class="row row-condensed"
    >
      <dl>
        <dt>
          Proposal ID
        </dt>
         
        <dd>
          2
        </dd>
      </dl>
       
      <dl>
        <dt>
          Submitted
        </dt>
         
        <dd>
          January 1st 1970, 00:00
        </dd>
      </dl>
       
      <dl>
        <dt>
          Voting Start Date
        </dt>
         
        <dd>
          January 3rd 1970, 00:00
        </dd>
      </dl>
       
      <dl>
        <dt>
          Voting End Date
        </dt>
         
        <dd>
          January 5th 1970, 00:00
        </dd>
      </dl>
    </div>
  </div>
   
  <modaldeposit-stub
    denom="STAKE"
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
   
  <modalvote-stub
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
</tmpage-stub>
`;

exports[`PageProposal should display proposal page should default tally to 0 if it's not yet present  1`] = `
<tmpage-stub
  class="small"
  data-title="Proposal"
  hideheader="true"
  subtitle=""
  title=""
>
  <div
    class="page-profile__header page-profile__section proposal"
  >
    <div
      class="page-profile__header__info"
    >
      <span
        class="proposal-status pink"
      >
        Voting Period
      </span>
       
      <div
        class="page-profile__status-and-title"
      >
        <h2
          class="proposal-title"
        >
          VotingPeriod proposal
        </h2>
      </div>
    </div>
     
    <div
      class="button-container"
    >
      <tmbtn-stub
        color="primary"
        disabled="true"
        id="deposit-btn"
        value="Deposit"
      />
       
      <tmbtn-stub
        color="primary"
        id="vote-btn"
        value="Vote"
      />
    </div>
     
    <div
      class="row"
    >
      <div
        class="column"
      >
        <dl>
          <textblock-stub
            content="custom text proposal description"
          />
        </dl>
      </div>
    </div>
  </div>
   
  <div
    class="page-profile__section"
  >
    <div
      class="row row-condensed"
    >
      <!---->
    </div>
     
    <div
      class="row row-condensed"
    >
      <dl>
        <dt>
          Total Vote Count
        </dt>
         
        <dd>
          0% / 0
        </dd>
      </dl>
       
      <dl>
        <dt>
          Yes
        </dt>
         
        <dd>
          
            0% /
            0
          
        </dd>
      </dl>
       
      <dl>
        <dt>
          No
        </dt>
         
        <dd>
          
            0% /
            0
          
        </dd>
      </dl>
       
      <dl>
        <dt>
          No with Veto
        </dt>
         
        <dd>
          
            0% /
            0
          
        </dd>
      </dl>
       
      <dl>
        <dt>
          Abstain
        </dt>
         
        <dd>
          
            0% /
            0
          
        </dd>
      </dl>
    </div>
     
    <div
      class="row row-condensed"
    >
      <dl>
        <dt>
          Proposal ID
        </dt>
         
        <dd>
          2
        </dd>
      </dl>
       
      <dl>
        <dt>
          Submitted
        </dt>
         
        <dd>
          January 1st 1970, 00:00
        </dd>
      </dl>
       
      <dl>
        <dt>
          Voting Start Date
        </dt>
         
        <dd>
          January 3rd 1970, 00:00
        </dd>
      </dl>
       
      <dl>
        <dt>
          Voting End Date
        </dt>
         
        <dd>
          January 5th 1970, 00:00
        </dd>
      </dl>
    </div>
  </div>
   
  <modaldeposit-stub
    denom="STAKE"
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
   
  <modalvote-stub
    proposalid="2"
    proposaltitle="VotingPeriod proposal"
  />
</tmpage-stub>
`;

exports[`PageProposal should display proposal page should show a loader if the necessary data hasen't been loaded 1`] = `
<tmpage-stub
  class="small"
  data-title="Proposal"
  hideheader="true"
  subtitle=""
  title=""
>
  <tmdataloading-stub
    title="Data is loadingâ€¦"
  />
</tmpage-stub>
`;

exports[`PageProposal shows an error if the proposal couldn't be found 1`] = `
<tmpage-stub
  class="small"
  data-title="Proposal"
  hideheader="true"
  subtitle=""
  title=""
>
  <tmdataerror-stub />
</tmpage-stub>
`;
